db/migrate/20240726083740_alter_articles_add_versioning.rb

```diff
@@ -0,0 +1,217 @@
+      t.string :unit, default: ''
+      t.string :note
+      t.boolean :availability, default: true, null: false
+      t.string :manufacturer
+      t.string :origin
+      t.string :order_number
+      t.datetime :updated_at
+    end
+
+    # copy all article fields into article_versions
+    articles = select_all('SELECT article_versions.id AS article_version_id, articles.* FROM article_versions JOIN articles ON articles.id = article_versions.article_id')
+    articles.each do |article|
+      update(%(
+        UPDATE article_versions SET
+          name = #{quote article['name']},
+          article_category_id = #{quote article['article_category_id']},
+          unit = #{quote article['unit']},
+          note = #{quote article['note']},
+          availability = #{quote article['availability']},
+          manufacturer = #{quote article['manufacturer']},
+          origin = #{quote article['origin']},
+          order_number = #{quote article['order_number']},
+          updated_at = #{quote article['updated_at']}
+        WHERE id = #{quote article['article_version_id']}
+      ))
+    end
+
+    remove_index :articles, %i[name supplier_id]
+
+    # drop article columns (now superfluous as they exist in article_versions):
+    change_table :articles do |t|
+      t.remove :name
+      t.remove :article_category_id
+      t.remove :unit
+      t.remove :note
+      t.remove :availability
+      t.remove :manufacturer
+      t.remove :origin
+      t.remove :order_number
+      t.remove :updated_at
+      t.remove :price
+      t.remove :tax
+      t.remove :deposit
+      t.remove :unit_quantity
+    end
+
+    # remove order_articles' reference to articles (reference now always goes through article_versions):
+    articles = select_all(%(
+      SELECT articles.id AS article_id, article_versions.id AS article_version_id
+      FROM articles
+      JOIN article_versions ON article_versions.article_id = articles.id
+      JOIN order_articles ON order_articles.article_id = articles.id
+      WHERE order_articles.article_version_id IS NULL
+      GROUP BY article_versions.article_id
+      ORDER BY article_versions.created_at DESC
+    ))
+
+    articles.each do |article|
+      update(%(
+        UPDATE order_articles
+        SET article_version_id = #{quote article['article_version_id']}
+        WHERE order_articles.article_version_id IS NULL
+          AND order_articles.article_id = #{quote article['article_id']}
+      ))
+    end
+
+    # Remove orphaned order articles (db inconsistencies due to lack of foreign key constraints):
+    delete('DELETE FROM order_articles WHERE order_articles.article_version_id IS NULL')
+
+    # De-duplicate article version (db inconsistencies due to lack of unique key for created_at and article_id):
+    duplicate_article_versions = select_all(%{
+      SELECT article_id, created_at
+      FROM article_versions
+      GROUP BY article_id, created_at
+      HAVING COUNT(*) > 1
```

Comment:
This would partially "fix" #1114:

```suggestion
      HAVING COUNT(*) > 1 AND article_id <> 0
```

However, as it's not really in the scope of #1058, I'm reluctant to add it. Maybe this should be rectified in a separate PR first - also see https://github.com/foodcoops/foodsoft/discussions/1119